### PR TITLE
cmake

### PR DESCRIPTION
**Pull request by andreaTP**
I hope this is appreciated...
just to help people to compile this on different architecture without diverging the provided Makefile and get helps from cmake reading OS configuration!

### CODE REVIEW

**Comment by andreaTP** [comment]
probably this will also help in #1

File: cuhe/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+FIND_PACKAGE(CUDA REQUIRED)
+
+INCLUDE(FindCUDA)
+
+FILE(GLOB SOURCES "*.cu" "*.h")
+
+LIST(APPEND CMAKE_CXX_FLAGS ${OpenMP_CXX_FLAGS})
+LIST(APPEND CMAKE_CXX_FLAGS "-m64 -std=c++11 -O3 -ffast-math -Wall")
+
+LIST(APPEND CUDA_NVCC_FLAGS "-ccbin gcc-${GCC_CUDA_VERISON} -std=c++11 -gencode arch=compute_${GPU_ARCH},code=sm_${GPU_ARCH}")
```
Comment (WeiDaiWD) [inline cuhe/CMakeLists.txt:20]: gcc-${GCC_CUDA_VERISON}
I'm wondering what value for GCC_CUDA_VERSION should I put there.
I got error: "No such file or directory".
I tried value "4.8.4" or "7.5" but neither of them work for me.

Comment (andreaTP) [inline cuhe/CMakeLists.txt:20]: I t depends on the version of gcc you will use within cuda probably you previously linked your version within your cuda installation (as I did having making this working).
Probably "4.8.4" will be available on "4.8" if this doesn't work please let me know the exact version of your libraries and I will try to repro the problem.

What do you think about using cmake?

**Comment by WeiDaiWD** [comment]
It works fine with "4.8". I cannot appreciate more for this.
Cmake is what I meant to use, to detect OS. But I have no experience before.
Is there a way to check GCC version and set it automatically, as well as the CUDA compute capability?

**Comment by andreaTP** [comment]
Unfortunately here we fall on a couple of issues:
gcc -> in most of the systems with cuda you need to specify a version that is not the system default, and so you can have any number of different gcc (i.e. 4.8 4.9) on your machine. Cuda ask for specific gcc versions each time and it will be a bet to find out what a user will like to use... 
compute capabilities -> here the problem that is not always true that your dev machine is the same on which you will run the code, moreover you can have multiple cuda devices so again I haven't found a reliable method to do that.

If you don't mind I will try also to add different tasks to cmake (static and dynamic library, JNI compatibility and so on...) please let me know how far I can go on your project! :-)

Can I bother privately you on the not yet published code of DHS (autocomplete and sorting especially)?

Thanks in advance!

**Comment by WeiDaiWD** [comment]
1. I see. Both situations apply to me now.
   I believe that Cmake is good enough for now. I can compile on OSX and Ubuntu quite easily.
   Do you mind if I take that Cmake file as a template to use in my other projects?
2. "Sorting" will be published soon. I'm just running a few tests on parameter selections.
   Unfortunately, "autocomplete" needs more time to be public. My coworkers and I need to get our paper accepted before we can release the code. I appreciate your interest and help in this project.

**Comment by andreaTP** [comment]
1. absolutely no problem, this is open source! :-) Glad to help in a so cool project!
   please squash line 17 and 18 in `LIST(APPEND CMAKE_CXX_FLAGS "${OpenMP_CXX_FLAGS} -m64 -std=c++11 -O3 -ffast-math -Wall")` because I noticed a problem there when generating the shared library.
2. not sure if you mind but  I can offer a bit of time for helping you on the sw engineering and in developing some kind of tools around your implementation, without any kind of credits...

**Comment by WeiDaiWD** [comment]
Of course you can develop any application. The library is published so that people can use it and improve it.
I'm actually struggling with the software design. The library should maintain its super fast performance, as described in our papers. As a consequence of that, it's now really hard to be user friendly or to be easily applied to other CUDA GPUs. So your help is appreciated. Thanks a lot!

**Pull request merged**